sdk/lib/cxxrt/guard.cc

```diff
@@ -71,9 +75,10 @@ namespace
 		 */
 		void unlock()
 		{
-			assert(high == LockBit);
-			high = 0;
-			futex_wake(&high, std::numeric_limits<uint32_t>::max());
+			Debug::Assert(high == LockBit, "Corrupt guard word");
+			high    = 0;
+			int res = futex_wake(&high, std::numeric_limits<uint32_t>::max());
+			Debug::Invariant(res == 0, "futex_wake failed; possible deadlock");
```

Comment:
I’m not sure I like having this as an invariant. If a compartment is single threaded or otherwise ensures that statics are not concurrently initialised, it’s fine for this to fail. In this function, we don’t know if that is the case.

Reply:
A) whoops, that should be `>= 0`, not `== 0`

B) `futex_wake` fails (`< 0`) only if the cap is invalid or there's not enough room on the (trusted or untrusted) stack for the cross-call.

Reply:
Right, case B is the one I'm thinking about.  It's fine for the futex wake to fail if there are no waiters.  We don't use a waiters flag, so we can't tell if that's the case.  In a lot of uses, this is used for lazy initialisation, but not for locking, so the failure is benign then.

Reply:
I suppose, but I think I'd rather trip an assertion in initialization than quietly fail to wake a thread after the system's flying?

Reply:
If it’s an assert, that’s fine, you can turn it off. If it’s an invariant, you get a crash in your compartment instead of working code.